test/controllers/goodsitems/AdditionalDocumentControllerSpec.scala

```diff
@@ -0,0 +1,178 @@
+import org.mockito.Mockito._
+import org.scalacheck.Arbitrary._
+import org.scalacheck.Gen._
+import org.scalatest.OptionValues
+import org.scalatest.mockito.MockitoSugar
+import org.scalatest.prop.PropertyChecks
+import play.api.data.Form
+import play.api.test.Helpers._
+import services.cachekeys.CacheKey
+import uk.gov.hmrc.customs.test.behaviours.{CustomsSpec, EndpointBehaviours}
+import uk.gov.hmrc.http.cache.client.CacheMap
+import uk.gov.hmrc.wco.dec.GovernmentAgencyGoodsItemAdditionalDocument
+import views.html.gov_agency_goods_items_add_docs
+
+import scala.concurrent.Future
+
+class AdditionalDocumentControllerSpec extends CustomsSpec
+  with PropertyChecks
+  with Generators
+  with OptionValues
+  with MockitoSugar
+  with EndpointBehaviours {
+
+  private def form = Form(govtAgencyGoodsItemAddDocMapping)
+
+  val additionalDocsPageUri = "/submit-declaration-goods/add-gov-agency-goods-items-additional-docs"
+
+  private def controller(user: Option[SignedInUser]) =
+    new AdditionalDocumentController(new FakeActions(user), mockCustomsCacheService)
+
+  val goodsItemGen = option(arbitrary[GovernmentAgencyGoodsItem])
+
+  def view(form: Form[GovernmentAgencyGoodsItemAdditionalDocument] = form, additionalDocs: Seq[GovernmentAgencyGoodsItemAdditionalDocument] = Seq()): String =
+    gov_agency_goods_items_add_docs(form, additionalDocs)(fakeRequest, messages, appConfig).body
+
+  ".onPageLoad" should {
+
+    behave like okEndpoint(additionalDocsPageUri)
+    behave like authenticatedEndpoint(additionalDocsPageUri)
+
+    "return OK" when {
+
+      "user is signed in" in {
+
+        forAll { user: SignedInUser =>
+
+          val result = controller(Some(user)).onPageLoad()(fakeRequest)
+
+          status(result) mustBe OK
+          contentAsString(result) mustBe view()
+        }
+      }
+    }
+
+    "return Unauthorized" when {
+
+      "user doesn't have an eori" in {
+
+        forAll { user: UnauthenticatedUser =>
+
+          val result = controller(Some(user.user)).onPageLoad(fakeRequest)
+
+          status(result) mustBe UNAUTHORIZED
+        }
+      }
+    }
+
+    "load data from cache" in {
+
+      forAll(arbitrary[SignedInUser], goodsItemGen) {
+        case (user, data) =>
+          withCleanCache(EORI(user.eori.value), CacheKey.goodsItem, data) {
+            val result = controller(Some(user)).onPageLoad(fakeRequest)
+
+            status(result) mustBe OK
+            contentAsString(result) mustBe view(additionalDocs = data.map(_.additionalDocuments).getOrElse(Seq.empty))
+          }
+      }
+    }
+  }
+
+  ".onSubmit" should {
+
+    behave like badRequestEndpoint(additionalDocsPageUri, POST)
+    behave like authenticatedEndpoint(additionalDocsPageUri, POST)
+
+    "return OK" when {
+      "user submits valid data" in {
+        forAll (arbitrary[SignedInUser],arbitrary[GovernmentAgencyGoodsItemAdditionalDocument], goodsItemGen) { case (user: SignedInUser, additionalDocument, goodsItem) =>
```

Comment:
Don't need a case here, you can just use :
```suggestion
        forAll (arbitrary[SignedInUser],arbitrary[GovernmentAgencyGoodsItemAdditionalDocument], goodsItemGen) { (user, additionalDocument, goodsItem) =>
```

Reply:
updated